### PR TITLE
Enable hpu_graph support for wav2vec2-asr + custom ctc_loss only for Gaudi2

### PR DESCRIPTION
**Pull request by vivekgoe**
# What does this PR do?

Fixes # (issue)
Enable hpu_graph support for wav2vec2-asr (add custom ctc_loss usage only for Gaudi2). Motivation to add hpu graphs is two-fold, (1) to improve performance, (2) to reduce run to run performance variations.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by yeonsily** [comment]
The change caused regression on 1.14 docker run. Can you please check it?

File: optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 1]
```diff
@@ -300,58 +369,84 @@ def gaudi_wav2vec2_encoder_forward(
...
+        input_lengths = self._get_feat_extract_output_lengths(attention_mask.sum(-1)).to(torch.long)
+        # assuming that padded tokens are filled with -100
+        # when not being attended to
+        labels_mask = labels >= 0
+        target_lengths = labels_mask.sum(-1)
+        # ctc_loss doesn't support fp16
+        log_probs = torch.nn.functional.log_softmax(logits, dim=-1, dtype=torch.float32).transpose(0, 1)
+        if get_device_name() == "GAUDI":
+            flattened_targets = labels.masked_select(labels_mask)
+            with torch.backends.cudnn.flags(enabled=False):
```
Comment (regisss) [inline optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py:426]: Do we need this on Gaudi?

File: optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 2]
```diff
@@ -300,58 +369,84 @@ def gaudi_wav2vec2_encoder_forward(
...
+                    flattened_targets,
+                    input_lengths,
+                    target_lengths,
+                    blank=self.config.pad_token_id,
+                    reduction=self.config.ctc_loss_reduction,
+                    zero_infinity=self.config.ctc_zero_infinity,
+                )
+        else:
+            flattened_targets = labels
+            with torch.backends.cudnn.flags(enabled=False):
```
Comment (regisss) [inline optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py:438]: Same

File: optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 3]
```diff
@@ -300,58 +369,84 @@ def gaudi_wav2vec2_encoder_forward(
...
     return_dict: Optional[bool] = None,
-) -> Union[Tuple, Wav2Vec2BaseModelOutput]:
+    labels: Optional[torch.Tensor] = None,
+) -> Union[Tuple, CausalLMOutput]:
+    r"""
+    labels (`torch.LongTensor` of shape `(batch_size, target_length)`, *optional*):
+        Labels for connectionist temporal classification. Note that `target_length` has to be smaller or equal to
+        the sequence length of the output logits. Indices are selected in `[-100, 0, ..., config.vocab_size - 1]`.
+        All labels set to `-100` are ignored (masked), the loss is only computed for labels in `[0, ...,
+        config.vocab_size - 1]`.
```
Comment (regisss) [inline optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py:389]: No need to keep this since it is exactly the same as in Transformers

File: optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 4]
```diff
@@ -300,58 +369,84 @@ def gaudi_wav2vec2_encoder_forward(
     )
 
 
-def gaudi_wav2vec2_forward(
+_HIDDEN_STATES_START_POSITION = 2
```
Comment (regisss) [inline optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py:372]: Maybe we can import `_HIDDEN_STATES_START_POSITION` from `transformers.models.wav2vec2.modeling_wav2vec2`?

**Comment by regisss** [comment]
> The change caused regression on 1.14 docker run. Can you please check it?

I think this happens because the CTC loss kernel is introduced in Synapse 1.15 right?
If it cannot work withg 1.14, we can merge this PR into the `synapse_1.15` branch.

File: optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 4 above)
Comment (vivekgoe) [reply optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py:372]: Ok, let me try doing that.

File: optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 3 above)
Comment (vivekgoe) [reply optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py:389]: Ok, will make the change.

File: optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 1 above)
Comment (vivekgoe) [reply optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py:426]: Custom ctc_loss operator works only for Gaudi2, therefore need this check.

File: optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 2 above)
Comment (vivekgoe) [reply optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py:438]: Same as above.

File: optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 1 above)
Comment (regisss) [reply optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py:419]: Yes yes this I get it, I was referring to:
```py
with torch.backends.cudnn.flags(enabled=False):
```
I don't think CuDNN is used anywhere on Gaudi no?

Comment (vivekgoe) [reply optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py:419]: We don't, I left it since it was already there. Will remove it.

Comment (regisss) [reply optimum/habana/transformers/models/wav2vec2/modeling_wav2vec2.py:419]: Thanks! I'm not sure how this can interact well or not with Habana Torch so I think it's always better to remove references to CUDA and similar GPU libraries.

**Comment by vivekgoe** [comment]
@regisss updated all review comments. @yeonsily @libinta updated code to ensure backward compatibility with Synapse 1.14.0.

File: examples/speech-recognition/README.md [hunk 5]
```diff
@@ -78,13 +78,15 @@ python run_speech_recognition_ctc.py \
...
+    --bf16 \
+    --use_hpu_graphs_for_training \
+    --use_hpu_grpahs_for_inference
 ```
 
 On a single HPU, this script should run in *ca.* 6 hours and yield a CTC loss of **0.059** and a word error rate of **0.0423**.
 
 > If your data has a sampling rate which is different from the one of the data the model was trained on, this script will raise an error.
-> Resampling with the `datasets` library is not supported on HPUs yet.
+> Resampling with the `datasets` library is not supported on HPUs yet. HPU graphs are supported only on Gaudi2.
```
Comment (regisss) [inline examples/speech-recognition/README.md:89]:
```suggestion
> Resampling with the `datasets` library is not supported on HPUs yet. HPU graphs are supported only on Gaudi2 and from SynapseAI v1.15.
```

File: examples/speech-recognition/README.md [hunk 6]
```diff
@@ -117,13 +119,15 @@ python ../gaudi_spawn.py \
...
+    --bf16 \
+    --use_hpu_graphs_for_training \
+    --use_hpu_graphs_for_inference
 ```
 
 On 8 HPUs, this script should run in *ca.* 49 minutes and yield a CTC loss of **0.0613** and a word error rate of **0.0458**.
 
 > If your data has a sampling rate which is different from the one of the data the model was trained on, this script will raise an error.
-> Resampling with the `datasets` library is not supported on HPUs yet.
+> Resampling with the `datasets` library is not supported on HPUs yet. HPU graphs are supported only on Gaudi2.
```
Comment (regisss) [inline examples/speech-recognition/README.md:130]:
```suggestion
> Resampling with the `datasets` library is not supported on HPUs yet. HPU graphs are supported only on Gaudi2 and from SynapseAI v1.15.
```

**Review by regisss — COMMENTED** [review]
I'm suggesting two small changes in the README because HPU graphs don't work on my Gaudi2 instance with 1.14.
